File: queue-api/src/main/java/net/mythicisland/queue/api/ticket/Ticket.java
package net.mythicisland.queue.api.ticket;

import net.mythicisland.queue.api.match.Assignment;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

/**
 * A single player or a group of players.
 *
 * @param id the unique ID of the ticket
 * @param playerIds the players behind the ticket, one entry means solo
 * @param queueTypes the queue types the ticket is searching in
 * @param state the current state of the ticket
 * @param createdAt when the ticket entered matchmaking
 * @param matchId the match the ticket was put into, null while searching
 * @param assignment the server to connect to, null until one was allocated
 * @param countdownEndsAt when the players get transferred, null until a server was allocated
 */
public record Ticket(
        UUID id,
        List<UUID> playerIds,
        List<String> queueTypes,
        TicketState state,
        Instant createdAt,
        UUID matchId,
        Assignment assignment,
        Instant countdownEndsAt
) {
}

File: queue-shared/src/main/kotlin/net/mythicisland/queue/shared/match/Ticket.kt
package net.mythicisland.queue.shared.match

import build.buf.gen.mythicisland.queue.v2.TicketState
import build.buf.gen.mythicisland.queue.v2.ticket
import net.mythicisland.common.util.protobuf.toTimestamp
import java.time.Instant
import java.util.UUID

data class Ticket(
    val id: UUID,
    val playerIds: List<UUID>,
    val queueTypes: List<String>,
    val state: TicketState,
    val createdAt: Instant,
    val matchId: UUID? = null,
    val assignment: Assignment? = null,
    val countdownEndsAt: Instant? = null,
) {

    fun toDefinition(): build.buf.gen.mythicisland.queue.v2.Ticket {
        return ticket {
            id = this@Ticket.id.toString()
            playerIds.addAll(this@Ticket.playerIds.map(UUID::toString))
            queueTypes.addAll(this@Ticket.queueTypes)
            state = this@Ticket.state
            createdAt = this@Ticket.createdAt.toTimestamp()
            this@Ticket.matchId?.let { matchId = it.toString() }
            this@Ticket.assignment?.let { assignment = it.toDefinition() }
            this@Ticket.countdownEndsAt?.let { countdownEndTime = it.toTimestamp() }
        }
    }

    fun asSearching(): Ticket {
        return copy(
            state = TicketState.TICKET_STATE_SEARCHING,
            matchId = null,
            assignment = null,
            countdownEndsAt = null,
        )
    }

}
